src/coordinator/ts/values.go

```diff
@@ -39,79 +38,98 @@ type Values interface {
 	// ValueAt returns the value at the nth element
 	ValueAt(n int) float64
 
-	// The number of millisseconds represented by each index
-	MillisPerStep() int
-
-	// Slice of data values in a range
-	Slice(begin, end int) Values
+	// DatapointAt returns the datapoint at the nth element
+	DatapointAt(n int) Datapoint
+}
 
-	// AllNaN returns true if the values are all NaN
-	AllNaN() bool
+// A Datapoint is a single data value reported at a given time
+type Datapoint struct {
+	Timestamp time.Time
+	Value     float64
 }
 
+// Datapoints is a list of datapoints.
+type Datapoints []Datapoint
+
+// Len is the length of the array.
+func (d Datapoints) Len() int { return len(d) }
+
+// ValueAt returns the value at the nth element.
+func (d Datapoints) ValueAt(n int) float64 { return d[n].Value }
+
+// DatapointAt returns the value at the nth element.
+func (d Datapoints) DatapointAt(n int) Datapoint { return d[n] }
+
 // MutableValues is the interface for values that can be updated
 type MutableValues interface {
 	Values
 
-	// Resets the values
-	Reset()
-
 	// Sets the value at the given entry
 	SetValueAt(n int, v float64)
 }
 
-// NewValues returns MutableValues supporting the given number of values at the
-// requested granularity.  The values start off as NaN
-func NewValues(ctx context.Context, millisPerStep, numSteps int) MutableValues {
-	return newValues(ctx, millisPerStep, numSteps, math.NaN())
+// FixedResolutionMutableValues are mutable values with fixed resolution between steps
+type FixedResolutionMutableValues interface {
+	MutableValues
+	Resolution() time.Duration
+	StepAtTime(t time.Time) int
+	StartTimeForStep(n int) time.Time
+	// Time when the series
+	StartTime() time.Time
+	MillisPerStep() int
 }
 
-func newValues(ctx context.Context, millisPerStep, numSteps int, initialValue float64) MutableValues {
-	values := make([]float64, numSteps)
+type fixedResolutionValues struct {
+	millisPerStep int
+	numSteps      int
+	values        []float64
+	startTime     time.Time
+}
 
-	// Faster way to initialize an array instead of a loop
-	Memset(values, initialValue)
-	vals := &float64Values{
-		ctx:           ctx,
-		millisPerStep: millisPerStep,
-		numSteps:      numSteps,
-		allNaN:        math.IsNaN(initialValue),
-		values:        values,
+func (b *fixedResolutionValues) MillisPerStep() int        { return b.millisPerStep }
+func (b *fixedResolutionValues) Len() int                  { return b.numSteps }
+func (b *fixedResolutionValues) ValueAt(point int) float64 { return b.values[point] }
```

Comment:
Won't this be 0 unless it's been explicitly set by `SetValueAt(_,_)`?

Reply:
yea it would be set to the initial value unless someone uses SetValueAt. 

We can chose to check check bounds in all these function calls but taking statsdex doesn't really do it and i think we can push that responsibility on the client.